src/ui/win/win_ui.cpp

```diff
@@ -0,0 +1,257 @@
+}
+
+UiWindowWin::~UiWindowWin()
+{
+    _free();
+}
+
+void UiWindowWin::_display()
+{
+    if ( !_shown ) // we don't need to display anything
+        return;
+
+    ShowWindow( _window, SW_SHOWDEFAULT );
+    UpdateWindow( _window );
+
+    MSG windowMessage;
+    while ( GetMessage( &windowMessage, NULL, 0, 0 ) ) {
+        TranslateMessage( &windowMessage );
+        DispatchMessage( &windowMessage );
+    }
+}
+
+void UiWindowWin::setImage( const PenguinV_Image::Image & image )
+{
+    _free();
+
+    // We need to set image upside-down to show correctly in UI window as well as make 4 byte row length for bitmap
+    _image.setAlignment( 4u );
+    _image.setColorCount( image.colorCount() );
+    _image.resize( image.width(), image.height() );
+
+    const uint32_t rowSizeIn  = image.rowSize();
+    const uint32_t rowSizeOut = _image.rowSize();
+    const uint32_t width      = image.width() * image.colorCount();
+
+    const uint8_t * inY     = image.data() + image.rowSize() * (image.height() - 1);
+    uint8_t       * outY    = _image.data();
+    const uint8_t * outYEnd = outY + image.height() * rowSizeOut;
+
+    for ( ; outY != outYEnd; outY += rowSizeOut, inY -= rowSizeIn )
+        memcpy( outY, inY, sizeof( uint8_t ) * width );
+
+    const bool rgbImage = (image.colorCount() != 1u);
+    const DWORD bmpInfoSize = sizeof( BITMAPINFOHEADER ) + (rgbImage ? 1 : 256) * sizeof( RGBQUAD );
+
+    _bmpInfo = reinterpret_cast<BITMAPINFO*>(malloc( bmpInfoSize ));
+    _bmpInfo->bmiHeader.biSize          = sizeof( BITMAPINFOHEADER );
+    _bmpInfo->bmiHeader.biWidth         = image.width();
+    _bmpInfo->bmiHeader.biHeight        = image.height();
+    _bmpInfo->bmiHeader.biPlanes        = 1;
+    _bmpInfo->bmiHeader.biBitCount      = image.colorCount() * 8;
+    _bmpInfo->bmiHeader.biCompression   = BI_RGB;
+    _bmpInfo->bmiHeader.biSizeImage     = image.rowSize() * image.height();
+    _bmpInfo->bmiHeader.biXPelsPerMeter = 0;
+    _bmpInfo->bmiHeader.biYPelsPerMeter = 0;
+    _bmpInfo->bmiHeader.biClrUsed       = rgbImage ? 0 : 255;
+    _bmpInfo->bmiHeader.biClrImportant  = 0;
+
+    if ( !rgbImage ) {
+        for ( uint8_t i = 0u; ; ++i ) {
+            _bmpInfo->bmiColors[i].rgbRed   = i;
+            _bmpInfo->bmiColors[i].rgbGreen = i;
+            _bmpInfo->bmiColors[i].rgbBlue  = i;
+            _bmpInfo->bmiColors[i].rgbReserved = 0;
+
+            if ( i == 255u ) // to avoid variable overflow
+                break;
+        }
+    }
+}
+
+void UiWindowWin::drawPoint( const Point2d & point, const PaintColor & color )
+{
+    _point.push_back( PointToDraw( point, color ) );
+
+    _display();
+}
+
+void UiWindowWin::drawLine( const Point2d & start, const Point2d & end, const PaintColor & color )
+{
+    _line.push_back( LineToDraw( start, end, color ) );
+
+    _display();
+}
+
+void UiWindowWin::_free()
+{
+    if ( _bmpInfo != nullptr ) {
+        free( _bmpInfo );
+        _bmpInfo = nullptr;
```

Comment:
Too defensive. No access to `_bmpinfo` after `free`.

Reply:
There is: `setImage()` calls this function. Of course it's done to prevent any future problems